x-pack/filebeat/input/cel/input.go

```diff
@@ -950,3 +950,104 @@ func newInputMetrics(id string) *inputMetrics {
 func (m *inputMetrics) Close() {
 	m.unregister()
 }
+
+// redactor implements lazy field redaction of sets of a mapstr.M.
+type redactor struct {
+	state  mapstr.M
+	mask   []string // mask is the set of dotted paths to redact from state.
+	delete bool     // if delete is true, delete redacted fields instead of showing a redaction.
+}
+
+// String renders the JSON corresponding to r.state after applying redaction
+// operations.
+func (r redactor) String() string {
```

Comment:
If you wanted the data to remain in a structured format and do it lazily you could implement https://pkg.go.dev/go.uber.org/zap@v1.24.0/zapcore#ObjectMarshalerFunc.MarshalLogObject. I don't have any preference.

Reply:
The recommended approach for doing that does not really simplify things from what I can see. This is from zap issue 750.